relays/ethereum/src/ethereum_client.rs

```diff
@@ -150,6 +214,172 @@ async fn transaction_receipt(client: Client, hash: H256) -> (Client, Result<Rece
 	)
 }
 
+/// Returns best Substrate block that PoA chain knows of.
+pub async fn best_substrate_block(
+	client: Client,
+	contract_address: Address,
+) -> (Client, Result<SubstrateHeaderId, Error>) {
+	let (encoded_call, call_decoder) = bridge_contract::functions::best_known_header::call();
+	let call_request = bail_on_arg_error!(
+		to_value(CallRequest {
+			to: Some(contract_address),
+			data: Some(encoded_call.into()),
+		})
+		.map_err(|e| Error::RequestSerialization(e)),
+		client
+	);
+	let (client, call_result) =
+		bail_on_error!(call_rpc::<Bytes>(client, "eth_call", Params::Array(vec![call_request]),).await);
+	let (number, raw_hash) = match call_decoder.decode(&call_result.0) {
+		Ok((raw_number, raw_hash)) => (raw_number, raw_hash),
+		Err(error) => return (client, Err(Error::ResponseParseFailed(format!("{}", error)))),
+	};
+	let hash = match SubstrateHash::decode(&mut &raw_hash[..]) {
+		Ok(hash) => hash,
+		Err(error) => return (client, Err(Error::ResponseParseFailed(format!("{}", error)))),
+	};
+
+	if number != number.low_u32().into() {
+		return (client, Err(Error::InvalidSubstrateBlockNumber));
+	}
+
+	(client, Ok(HeaderId(number.low_u32(), hash)))
+}
+
+/// Returns true if Substrate header is known to Ethereum node.
+pub async fn substrate_header_known(
+	client: Client,
+	contract_address: Address,
+	id: SubstrateHeaderId,
+) -> (Client, Result<(SubstrateHeaderId, bool), Error>) {
+	let (encoded_call, call_decoder) = bridge_contract::functions::is_known_header::call(id.1);
+	let call_request = bail_on_arg_error!(
+		to_value(CallRequest {
+			to: Some(contract_address),
+			data: Some(encoded_call.into()),
+		})
+		.map_err(|e| Error::RequestSerialization(e)),
+		client
+	);
+	let (client, call_result) =
+		bail_on_error!(call_rpc::<Bytes>(client, "eth_call", Params::Array(vec![call_request]),).await);
+	match call_decoder.decode(&call_result.0) {
+		Ok(is_known_block) => (client, Ok((id, is_known_block))),
+		Err(error) => (client, Err(Error::ResponseParseFailed(format!("{}", error)))),
+	}
+}
+
+/// Submits Substrate headers to Ethereum contract.
+pub async fn submit_substrate_headers(
+	client: Client,
+	params: EthereumSigningParams,
+	contract_address: Address,
+	headers: Vec<QueuedSubstrateHeader>,
+) -> (Client, Result<(Vec<H256>, Vec<SubstrateHeaderId>), Error>) {
```

Comment:
What do you think about:

```
type EthereumTransactionHash = H256;
```

and using that here (and all the functions that follow)?